### ISSUE TITLE
Bug in CViewContainer::removeAll?

### ISSUE THREAD

**Issue by HolgerStrauss** [steinbergmedia/vstgui#40]
Hi,
I may have found a bug in CViewContainer::removeAll:
```
bool CViewContainer::removeAll (bool withForget)
{
	if (pImpl->mouseDownView)
		pImpl->mouseDownView = nullptr;
	pImpl->currentDragView = nullptr;
	
	ViewList::iterator it = pImpl->children.begin ();
	while (it != pImpl->children.end ())
	{
		CView* view = *it;
		if (isAttached ())
			view->removed (this);
		pImpl->children.erase (it);
		view->setSubviewState (false);
		pImpl->viewContainerListeners.forEach ([&] (IViewContainerListener* listener) {
			listener->viewContainerViewRemoved (this, view);
		});
		if (withForget)
			view->forget ();
		it = pImpl->children.begin ();
	}
	return true;
}
```
The _children_ member is actually of type _ViewList_, which is defined as _std::list<SharedPointer<CView>>_. 
If all SharedPointer's to a _CView_ have  been destroyed already, except for a single one left in CViewContainer's _children_ member, the line
`		pImpl->children.erase (it);
`
will call the CView's destructor and the object will be destroyed.
However, a pointer to this object is still in the _view_ variable and used later on, e.g. `view->setSubviewState (false);`. As far as I can see, this is an illegal access to a destroyed object.
It may probably be solved by moving `pImpl->children.erase (it);` further down in the while loop.

I can create a pull request of this change, but I'd like to hear an expert's option before doing so.

**Comment by scheffle** [comment]
Hi Holger,
the earliest destruction of the view will be with the call to view->forget (). Afterwards the view pointer is not used anymore. Only a leak may happen, if you call this method with 'withForget==false'.
Cheers
Arne

**Comment by HolgerStrauss** [comment]
Arne, thanks for your quick response.
I've re-checked my code in the debugger and I've come to the conclusion, that I am probably using VSTGUI in an "unintended" may.
Calling removeAll with 'withForget==true' will actually decrement the reference count of the CView objects twice: The first time, the erase on the _children_ container will call the SharedPointer's destuctor, which calls _forget_. The second time explicitly because of 'withForget = true'. For this to work, objects in the CViewContainer must have a reference count of at least 2. However, adding a CView to a CViewContainer using CViewContainer::addView only increments the reference count of the added object by one. So one must make sure, that the code that puts the control into the container actually has a "logical reference count leak" by 1, e.g.
```
auto p = new VSTGUI::CTextEdit(...);
pViewContainer->addView(p);
```
So the constructor of VSTGUI::CTextEdit sets the reference count to 1 and after calling addView, it will be 2. 

The mistake that I made is to avoid raw pointers (which is recommended by modern C++ style guidelines), so I used:
```
auto p = VSTGUI::makeOwned<CTextEdit>(...)
pViewContainer->addView(p);

```
This way, the constructing code doesn't actually have a "logical reference count leak". However, the CView in the CViewContainer will have a reference count of 1 after p goes out of scope. To cope for this, I was calling CViewContainer::removeAll with 'withFormat = false'. And then the issue described in my post happens. Probably, it should not be done this way and I know a way to work around it now.

Nevertheless, would you agree that the change proposed in my inital post will allow to handle both "use cases" at the same time?

**Comment by scheffle** [comment]
Your proposal has one little issue, it can change the runtime behaviour, because the listeners are then called while the view is still in the list. I would propose a different change, which will allow both "use-cases" : Change the line :
`CView* view = *it;`
with
`auto view = shared (*it);`
This just makes sure that the view is not destroyed before changing the scope.
How is this ?

**Comment by HolgerStrauss** [comment]
I understand your point.
`auto view = shared(*it);
`
does not compile (error C2672: 'shared' no matching overload function found).
But
`auto view = *it;
`
will do the trick, because view is a copy of the SharedPointer then.
If you agree, I'd be happy to create a pull request, but I won't be able to do it before Friday.
Cheers,
Holger

**Comment by scheffle** [comment]
Yeah, that would be nice. Thanks.

**Comment by scheffle** [comment]
Issue solved with #42
